crates/runtime/src/instance/allocator/pooling/index_allocator.rs

```diff
@@ -285,150 +205,197 @@ impl IndexAllocator {
-                        per_module.remove(&id.unwrap());
-                    }
-                    // Make sure to remove from the global
-                    // freelist. We already removed from the
-                    // per-module list above.
-                    remove_global_free_list_item(slot_state, free_list, slot_id);
-                    slot_state[slot_id.index()] = SlotState::Taken(id);
-                    Some(slot_id)
-                } else {
-                    if free_list.len() == 0 {
-                        return None;
-                    }
-                    // Pick a random free slot ID. Note that we do
-                    // this, rather than pick a victim module first,
-                    // to maintain an unbiased stealing distribution:
-                    // we want the likelihood of our taking a slot
-                    // from some other module's freelist to be
-                    // proportional to that module's freelist
-                    // length. Or in other words, every *slot* should
-                    // be equally likely to be stolen. The
-                    // alternative, where we pick the victim module
-                    // freelist first, means that either a module with
-                    // an affinity freelist of one slot has the same
-                    // chances of losing that slot as one with a
-                    // hundred slots; or else we need a weighted
-                    // random choice among modules, which is just as
-                    // complex as this process.
-                    //
-                    // We don't bother picking an empty slot (no
-                    // established affinity) before a random slot,
-                    // because this is more complex, and in the steady
-                    // state, all slots will see at least one
-                    // instantiation very quickly, so there will never
-                    // (past an initial phase) be a slot with no
-                    // affinity.
-                    let free_list_index = inner.rng.gen_range(0..free_list.len());
-                    let slot_id = free_list[free_list_index];
-                    // Remove from both the global freelist and
-                    // per-module freelist, if any.
-                    remove_global_free_list_item(slot_state, free_list, slot_id);
-                    if let &SlotState::Free(FreeSlotState::Affinity { module, .. }) =
-                        &slot_state[slot_id.index()]
-                    {
-                        remove_module_free_list_item(slot_state, per_module, module, slot_id);
-                    }
-                    slot_state[slot_id.index()] = SlotState::Taken(id);
 
-                    Some(slot_id)
-                }
+        let slot_id = match inner.strategy {
+            PoolingAllocationStrategy::NextAvailable => *inner.free_list.last()?,
+            PoolingAllocationStrategy::Random => inner.alloc_random()?,
```

Comment:
something about these two lines together feels confusing to me (though they are correct): the NextAvailable case picks but does not remove the index we're going to allocate, while `alloc_random` sounds like it allocates (i.e. removes from freelist).

Perhaps can we (i) add a comment before these two that we're going to remove from the freelist below, and (ii) rename the `alloc_*` helpers to `pick_*`, and (iii) use a helper for the first case here too?